src/box/sql/build.c

```diff
@@ -1940,7 +1865,7 @@ sql_create_foreign_key(struct Parse *parse_context)
 		 * If parent columns are not specified, then PK
 		 * columns of parent table are used as referenced.
 		 */
-		struct index *parent_pk = space_index(parent_space, 0);
+		const struct index *parent_pk = parent_space->index[0];
```

Comment:
Oops, this change is illegal. Fortunately, `space_index` checks whether `space->index` is allocated (i.e. `space->index_id_max` is greater than zero, so space has an index). Consider the following example from our tests.

```
tarantool> box.schema.space.create('t1')
tarantool> box.execute([[CREATE TABLE t (i INT PRIMARY KEY, a INT REFERENCES t1);]])
```

As you can see `t1` doesn't have indices, and ASan consider fetching 0th item as a boundaries violation. Hence, I suggest to revert this particular hunk, but consider `const` qualifier for `parent_space`.

Reply:
Here is my suggestion for fixup.

```diff
diff --git a/src/box/space.h b/src/box/space.h
index 604009ce1a..e60e1e1d05 100644
--- a/src/box/space.h
+++ b/src/box/space.h
@@ -430,7 +430,7 @@ space_run_triggers(struct space *space, bool yesno);
  * @return NULL if the index is not found.
  */
 static inline struct index *
-space_index(struct space *space, uint32_t id)
+space_index(const struct space *space, uint32_t id)
 {
        if (id <= space->index_id_max)
                return space->index_map[id];
diff --git a/src/box/sql/build.c b/src/box/sql/build.c
index 01016fd328..f81fa1765f 100644
--- a/src/box/sql/build.c
+++ b/src/box/sql/build.c
@@ -1865,7 +1865,7 @@ sql_create_foreign_key(struct Parse *parse_context)
                 * If parent columns are not specified, then PK
                 * columns of parent table are used as referenced.
                 */
-               const struct index *parent_pk = parent_space->index[0];
+               const struct index *parent_pk = space_index(parent_space, 0);
                if (parent_pk == NULL) {
                        diag_set(ClientError, ER_CREATE_FK_CONSTRAINT,
                                 constraint_name,

```

Reply:
I think that would be overkill. We can simply check that index_count is 0 and throw an error if it is. Otherwise, index[0] exists and is the PK.